lib/bloom/sbbf.ts

```diff
@@ -0,0 +1,408 @@
+
+        m = (m + SplitBlockBloomFilter.NUMBER_OF_BLOCKS - 1) & (~SplitBlockBloomFilter.NUMBER_OF_BLOCKS);
+
+        // Handle overflow:
+        const upperShiftL3 = SplitBlockBloomFilter.UPPER_BOUND_BYTES << 3
+        if (m > upperShiftL3 || m < 0 ) {
+            m = upperShiftL3;
+        }
+
+        // Round numBits up
+        m = (m + SplitBlockBloomFilter.BITS_PER_BLOCK -1) & ~SplitBlockBloomFilter.BITS_PER_BLOCK
+
+        const lowerBoundShiftL3 = SplitBlockBloomFilter.LOWER_BOUND_BYTES << 3
+        if (m < lowerBoundShiftL3 ) {
+            m = lowerBoundShiftL3;
+        }
+
+        return Math.ceil(m / this.BITS_PER_BLOCK)
+    }
+
+    /**
+     * @function mask
+     * @description generate a mask block for a bloom filter block
+     * @param hashValue: the hash value to generate the mask from
+     * @private
+     *
+     * @return mask Block
+     */
+    static mask(hashValue: Long): Block {
+        let result: Block = SplitBlockBloomFilter.initBlock()
+        for (let i = 0; i < result.length; i++) {
+            const y = hashValue.getLowBitsUnsigned() * SplitBlockBloomFilter.salt[i]
+            result[i] = result[i] | (1 << (y >>> 27))
+        }
+        return result
+    }
+
+    /**
+     * @function blockInsert
+     * @description insert a hash into a Bloom filter Block
+     * @param b: the block to flip a bit for: is changed
+     * @param hashValue: the hash value to insert into b
+     * @private
+     *
+     * @return void
+     */
+    // FIXME: hashValue should be a Long val  ?
+    // TODO: Make sure the value is preserved; I think it's passed by ref
+    static blockInsert(b: Block, hashValue: Long): void {
+        const masked: Block = this.mask(hashValue)
+        for (let i = 0; i < masked.length; i++) {
+            for (let j = 0; j < this.WORD_SIZE; j++) {
+                const isSet = masked[i] & (2 ** j)
+                if (isSet) {
+                    b[i] = b[i] | (2 ** j)
+                }
+            }
+        }
+    }
+
+    /**
+     * @function blockCheck
+     * @description check if a hashValue exists for this filter
+     * @param b: the block to check for inclusion
+     * @param hashValue: the hash to check for  should be long
+     * @private
+     *
+     * @return true if hashed item is __probably__ in the data set represented by this filter
+     * @return false if it is __definitely not__ in the data set.
+     */
+    // FIXME: hashValue should be a Long val   ?
+    static blockCheck(b: Block, hashValue: Long): boolean {
+        const masked: Block = this.mask(hashValue)
+        for (let i = 0; i < masked.length; i++) {
+            for (let j = 0; j < this.WORD_SIZE; j++) {
+                const isSet = masked[i] & (2 ** j)
+                if (isSet) {
+                    const match = b[i] & (2 ** j)
+                    if (!match) {
+                        return false
+                    }
+                }
+            }
+        }
```

Comment:
This feels wrong. Why are we digging in so deep to do the checking? Shouldn't we be able to do a bitwise mask, but after looking around, it appears that the answer is no.

Reply:
Correct. This is the specified way to do this algorithm.

Reply:
@shannonwells Not to change (as we want the standard way), but I thought of this last night. I think it works? (It passes the tests, but I'm not sure how well tested it is.)

```js
static blockCheck(b: Block, hashValue: Long): boolean {
  // This should likely be a constant like this.WORD_SIZE without my hacked mask creation.
  const WORD_MASK = parseInt(Array(this.WORD_SIZE).fill("1").join(""), 2);
  const masked: Block = this.mask(hashValue)
  for (let i = 0; i < masked.length; i++) {
    const hashMasked = masked[i] & WORD_MASK;
    if ((hashMasked & b[i]) != hashMasked) return false;
  }
  return true
}
```

Reply:
I'd rather not change the algorithm from the prescribed method without massive testing and validation.